modules/core/src/main/java/org/apache/ignite/internal/processors/affinity/GridAffinityAssignmentCache.java

```diff
@@ -304,6 +309,60 @@ public void onReconnected() {
     }
 
     /**
+     * @param assignment List indexed by partition number.
+     * @param nodesCnt Nodes count.
+     */
+    private void printDistribution(List<List<ClusterNode>> assignment, int nodesCnt) {
+        Float ignitePartDistribution = getFloat(IGNITE_PART_DISTRIBUTION_WARN_THRESHOLD, 0.1f);
+
+        int[] partitionsByLocalNode = new int[nodesCnt];
+
+        int totalPrimaryCnt = 0;
+
+        int totalBackupCnt = 0;
+
+        for (int i = 0; i < nodesCnt; i++) {
+            for (List<ClusterNode> partitionByNodes : assignment) {
+                if (partitionByNodes.get(i).isLocal())
+                    partitionsByLocalNode[i] += 1;
+
+                if (i == 0)
+                    totalPrimaryCnt++;
+                else
+                    totalBackupCnt++;
+            }
+        }
+
+        if (partitionsByLocalNode.length != 0) {
+            int localPrimaryCnt = partitionsByLocalNode[0];
+            int localBackupCnt = 0;
+            float localBackupPercent = 0;
+
+            for (int i = 1; i < partitionsByLocalNode.length; i++) {
+                localBackupCnt += partitionsByLocalNode[i];
+                localBackupPercent += (float)partitionsByLocalNode[i] / assignment.size() * 100;
+            }
+
+            float expectedCnt = (float)assignment.size() / nodesCnt;
+            float expectedPercent = expectedCnt / assignment.size() * 100;
+            float primaryPerNodePercent = (float)localPrimaryCnt / assignment.size() * 100;
+
+            float deltaPrimary = Math.abs(1 - (float)localPrimaryCnt * nodesCnt / totalPrimaryCnt);
+            float deltaBackup = Math.abs(1 - (float)localBackupCnt * nodesCnt / totalBackupCnt);
+
+            if (deltaPrimary > ignitePartDistribution || deltaBackup > ignitePartDistribution) {
+                log.info("Local node affinity assignment distribution is not ideal [cache=" + cacheOrGrpName +
```

Comment:
OK I added it 
String.format("Local node affinity assignment distribution is not ideal " + 
`"[cache=%s, expectedPrimary=%.2f](%.2f%%) ....]", cacheOrGrpName, expectedCnt, expectedPercent, ...) );`